rfd/0139-gogo-must-go.md

```diff
@@ -0,0 +1,132 @@
+
+#### CrowdStrike's `csproto` library
+
+https://github.com/CrowdStrike/csproto
+
+CrowdStrike used `gogo/protobuf` to overcome various problems with Google's original protobuf
+library and subsequently ran into issues during `gogo/protobuf`'s deprecated. `csproto`
+is a library that elects to use the proper protobuf runtime for the appropriate message.
+As a result, you could have an intermingling of different protobuf libraries without compatibility
+issues.
+
+### Our complications
+
+Teleport has a number of complications tied in with our use of `gogo/protobuf` that will make
+our migration unique. Along with our complications will be listed a number of potential
+mitigation strategies to discuss.
+
+**The primary issue here is our use of `gogo/protobuf` extensions.** We should strive
+to remove these extensions while maintaining backwards compatibility.
+
+#### JSON/YAML (de)serialization (removing `gogoproto.jsontag`)
+
+We rely heavily on `gogo/protobuf`'s custom JSON tags for serialization our messages into JSON.
+Additionally, we use a library called [`ghodss/yaml`](https://github.com/ghodss/yaml) to create YAML,
+which also relies on these JSON tags.
+
+Each resource is serialized into JSON before being pushed into our backend, which means the
+protobuf JSON tags have an impact quite far down our stack. There are a number of potential
+mitigations here with differing levels of impact. The primary goal of any mitigation strategy
+here should be **migrate away from gogoproto.jsontag extensions.**
+
+##### Serialize objects in backend as the gRPC wire format
+
+Serializing the objects in the backend as the gRPC wire format has a number of benefits:
+
+- Seems to remove the need for most unmarshaling functions in `lib/services`. (confirm?)
+- The backend has no need for JSON marshaling/unmarshaling, so JSON changes will not
+  affect the backend.
+- Allows us to remove gogoproto extensions without affecting backend storage.
+- Allows us to utilize fields in the backend more efficiently and store larger objects.
+
+The primary disadvantage here is that manually examining the backend is more difficult as
+the stored values will no longer be human readable.
+
+##### Write custom marshalers for objects
+
+In order to keep our user facing representations of the objects, we will need to write
+custom JSON and YAML marshalers for our objects. This may be time intensive but not overall
+a significant amount of work.
+
+#### Casting types (removing `gogoproto.casttype`)
+
+We additionally rely heavily on `gogoproto.casttype`, which allows us to avoid doing explicit
+casts or conversions within go code. We will have to take on this conversion work, unfortunately.
+
+#### Other gogoproto tags
+
+##### `gogoproto.nullable`
+
+We will need to remove the `gogoproto.nullable` tags, which overall shouldn't be a huge impact.
+We should migrate to using `optional` as defined [here](https://protobuf.dev/programming-guides/proto3/#field-labels).
```

Comment:
I know support for `optional` hasn't been in `proto3` since initiation - and I've not used it yet with `proto3`. If we can't use it, then we can use the google provided wrappers: https://github.com/protocolbuffers/protobuf/blob/main/src/google/protobuf/wrappers.proto

Reply:
Nullable tags are actually client side only and they are null if they are not set in proto. This happens when the value is the default value for the type so using a wrapper isn't an option as well because they change the wire representation

If a int field has zero value, it won't be marshalled to proto

Reply:
Oh yuck, I didn't even think about the fact that `optional` changes the wire representation. Yeah, this won't work.

Reply:
Well, all the same, `CheckAndSetDefaults` will realistically handle this for most structures. I think we can probably get away with just removing nullable.

Reply:
Yeah, nullable has no effect so we can handle internally

Reply:
I think y'all got this backwards - `optional` is the default for all proto3 fields. It is `required` that got dropped between proto2 and proto3, because as it turns out `required` was a terrible idea to begin with.

I don't recommend writing `optional` before our proto fields, and strongly not recommend messing with `required` (since I'm mentioning it). None of this changes the wire format - that is something that just doesn't happen with Google protobuf.

Reply:
It's not as simple as it is on the tin. You are correct that `proto3` gets rid of `required` and leans towards what could be called `optional` fields. However, an `optional` keyword was introduced in proto v3.15. With the Go generator, this usually yields a pointer to the expected value w/ something like:

```proto
message MyMessage {
  optional string foo = 1;
}
```

```go
type MyMessage struct {
  // various pb fields snipped
  Foo *string
}
```

I can't say I'm a huge fan of pointers being used to indicate the presence/non-presence of a field.

Reply:
I'd hoped we learned our lesson about custom generator options by now.

Reply:
`optional` is a perfectly legitimate feature (especially on scalar fields where you wouldn't otherwise get any "presence" info), but it's not supported by the javascript/typescript generator we're using, and gogoproto makes _terrible_ code for it.

Reply:
My usual advice would be to not rely on field presence - ideally you shouldn't care whether it's empty, nil, set explicitly, a default value, etc. If you do care about presence, then it's best to design in a way that makes it explicitly clear. A good example is using a FieldMask to inform partial updates - https://cloud.google.com/apis/design/standard_methods#update.